sbncode/SBNEventWeight/App/SystToolsEventWeight_module.cc

```diff
@@ -0,0 +1,238 @@
+
+        for( systtools::event_unit_response_t::iterator it = resp.begin(); it != resp.end(); ++it ){
+
+          //==== responses.size() is the number of universes
+          systtools::SystParamHeader const &sph = fParamHeaderHelper.GetHeader( (*it).pid );
+          std::string prettyName = sph.prettyName;
+
+          if(fDebugMode){
+            std::cout << "[SystToolsEventWeight::produce]       pid of this resp = " << (*it).pid << "\n"
+                      << "[SystToolsEventWeight::produce]       prettyName of this resp = " << prettyName << "\n"
+                      << "[SystToolsEventWeight::produce]       paramVariations.size() of this resp (expected) = " << sph.paramVariations.size() << "\n"
+                      << "[SystToolsEventWeight::produce]       responses.size() of this resp (produced) = " << (*it).responses.size() << std::endl;
+          }
+          if(sph.paramVariations.size()!=(*it).responses.size()){
+            std::cerr << "[SystToolsEventWeight::produce]       prettyName of this resp = " << prettyName << std::endl;
+            std::cerr << "[SystToolsEventWeight::produce]       We expect to have " << sph.paramVariations.size() << " universes, but "
+                      << (*it).responses.size() << " are produced. "
+                      << "Probably this particular event is not relevant to this systematic variation." << std::endl;
+            throw std::exception();
+          }
+
+          std::vector<float> wgts;
+          for( unsigned int i = 0; i < (*it).responses.size(); ++i ){
+
+            if(fDebugMode) std::cout << "[SystToolsEventWeight::produce]         (*it).responses[i] = " << (*it).responses[i] << std::endl;
+            wgts.push_back( (*it).responses[i] );
+
+          } // END loop over universes of this knob
+
+          mcwgh[sp->GetFullyQualifiedName()+"_"+prettyName] = wgts;
+
+        } // END loop over knobs
+
+        mcwghvec->push_back(std::move(mcwgh));
+        art::Ptr<sbn::evwgh::EventWeightMap> wghPtr = makeWeightPtr(mcwghvec->size() - 1);
+        wghassns->addSingle(mclist.at(nMCTruthIndex), wghPtr);
+
+        nMCTruthIndex++;
+
+      } // END loop over MCTruth particles
+
+    } // END systprovider loop
+
+  }
+
+  e.put(std::move(mcwghvec));
+  e.put(std::move(wghassns));
+
+}
+
+
+void SystToolsEventWeight::beginRun(art::Run& run) {
+
+  auto p = std::make_unique<std::vector<sbn::evwgh::EventWeightParameterSet> >();
+  for( auto &sp : fSystProviders ) {
+    MF_LOG_INFO("SystToolsEventWeight") << "sp->GetToolType() = " << sp->GetToolType() << "\n"
+                                  << "sp->GetFullyQualifiedName() = " << sp->GetFullyQualifiedName() << "\n"
+                                  << "sp->GetInstanceName() = " << sp->GetInstanceName() << "\n"
+                                  << "Printing each SystParamHeader of this ISystProviderTool..";
+    //==== Note: typedef std::vector<SystParamHeader> SystMetaData;
+    auto smd = sp->GetSystMetaData();
```

Comment:
Avoid a potential copy, and make sure you don't mistakenly attempt to change the values in there:
```suggestion
    auto const& smd = sp->GetSystMetaData();
```

Reply:
Updated as suggested.